New class QgsJSONUtils with utilities for converting features to and from GeoJSON strings

nyalldawson opened the pull request:
Title says it all!

nyalldawson committed:
[FEATURE] New class QgsJSONUtils with utilities for converting
features to and from GeoJSON strings

Sponsored by Kanton of Zug, Switzerland

nyalldawson committed:
Blacklist PyQgsJSONUtils test from qt5 (uses null)

m-kuhn commented:
:stuck_out_tongue_closed_eyes: e233b0b

nice stuff. did you checkout [qt5 json support](http://doc.qt.io/qt-5/json.html)? Do you think that can be valuable to us in the future?

nyalldawson commented:
@m-kuhn yes, it sounds very nice! For 3.0 I'll add methods which return the Qt5 json objects too, but given that most of the time the desired end result will usually be a string anyway, the methods implemented here will still be useful/required.

rouault commented on src/core/qgsjsonutils.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    default:
+    case QVariant::String:
+      QString v = value.toString().replace( '"', "\\\"" )
```
Looking at http://json.org/ , there are other characters ( \ \b \f \t / ) that would need to be \ escaped. See also https://github.com/json-c/json-c/blob/master/json_object.c#L109

nyalldawson replied:
Thanks @rouault! I'll fix this

tomchadwin commented:
Literally just saw this:

 https://mailarchive.ietf.org/arch/msg/geojson/B5lKKb6mSf2YHAiJJXDfHs_uv-8

If adopted, do we import/export only GeoJSON v2?

nyalldawson commented:
@tomchadwin hmmm... This current implementation ignores the crs and does not include any crs property. 

I'd be tempted to just say that QgsJSONUtils::featureToGeoJSON will ALWAYS expect geometries in WGS84. That way we'd already be compatible with GeoJSON v2. Possibly we could make featureToGeoJSON accept an optional coordinate transform (ie from the feature's native CRS to WGS84) and handle this conversion automatically if the transform is specified too.

rldhont commented:
If we don't want to transform and add CRS information ?

tomchadwin commented:
GeoJSON = EPSG:4326 has been an informal convention for a good while, so I think for export we are probably fine. But the link above suggests we should detect other CRS info and interpret as v1 if present, and therefore transform.

nyalldawson committed:
Correctly escape more json strings

nyalldawson committed:
Support encoding lists and maps to json

nyalldawson committed:
Refactor to move JSON exporter to its own class

nyalldawson committed:
Allow injecting extra properties into feature's GeoJSON export

Sponsored by Kanton of Zug, Switzerland

nyalldawson commented:
I've added a bunch of new stuff:
- support for encoding lists and maps in json output
- I've moved the GeoJSON exporter to its own class (the interface was become to unwieldy for a single function)
- Extra properties can now be injected into the GeoJSON feature's properties

tomchadwin commented:
> Extra properties can now be injected into the GeoJSON feature's properties

That could be a great feature. How about the converse - not including specified properties so that they are omitted from the GeoJSON?

nyalldawson committed:
Allow setting list of attributes to exclude from json exports

nyalldawson commented:
> That could be a great feature. How about the converse - not including specified properties so that they are omitted from the GeoJSON?

Ok - done. You could previously set a list of attributes to include manually, but I've added methods to set a list of attributes to specifically exclude too.

nyalldawson committed:
Support including attributes of related (child) features in
GeoJSON feature export

Sponsored by Kanton of Zug, Switzerland

nyalldawson committed:
Add method for exporting QgsFeatureList to GeoJSON featurecollection

nyalldawson committed:
Always create valid GeoJSON (both geometry and properties members
are required

nyalldawson committed:
Always export GeoJSON features in WGS84 (match specifications)

nyalldawson committed:
[FEATURE] Allow copying features from QGIS in GeoJSON format

The previous setting for include WKT when copying features has been
replaced with a choice of copying features as "Plain text, attributes
only", "Plain text, WKT geometry" and a new "GeoJSON" option.

When set to "GeoJSON", copying features in QGIS will place a GeoJSON
text representation of the features on the clipboard for easy
pasting into other applications/javascript code.

Sponsored by North Road

nyalldawson committed:
[server] Always export GeoJSON in WGS84 (fix #14609)

nyalldawson committed:
Catch errors when reprojecting for GeoJSON export

nyalldawson commented:
@rldhont The GeoJSON export is now always reprojected to WGS84 - see 8f5757f and f0b0852. That should resolve http://hub.qgis.org/issues/14609

nyalldawson commented:
Ok, last part I had planned for this just added - now you can choose to copy features from QGIS as GeoJSON FeatureCollections.

nyalldawson committed:
[FEATURE] Embed atlas feature into composer HTML source as GeoJSON

This change makes the current atlas feature (and additionally all
attributes of related child features) available to the source of
a composer HTML item, allowing the item to dynamically adjust
its rendered HTML in response to the feature's properties. An
example use case is dynamically populating a HTML table with
all the attributes of related child features for the atlas
feature.

To use this, the HTML source must implement a "setFeature(feature)"
JavaScript function. This function is called whenever the atlas
feature changes, and is passed the atlas feature (+related attributes)
as a GeoJSON Feature.

Sponsored by Kanton of Zug, Switzerland

nyalldawson commented:
Manually merged

tomchadwin commented:
Splendid work, Nyall.
